sdks/python/apache_beam/runners/worker/opcounters.py

```diff
@@ -24,11 +24,118 @@
 import math
 import random
 
+from apache_beam.utils import counters
 from apache_beam.utils.counters import Counter
+from apache_beam.utils.counters import CounterName
 
 # This module is experimental. No backwards-compatibility guarantees.
 
 
+class TransformIOCounter(object):
+  """Class to track time and bytes consumed while reading from IO.
+
+  Subclasses should be able to track consumption of IO across steps
+  in the same stage - for instance, if a Shuffle or Side Input iterable
+  is passed down to a next step.
+
+  Some examples of IO can be side inputs, shuffle, or streaming state.
+  """
+
+  def check_step(self):
+    """Check the current step within a stage as it may have changed.
+
+    If the state changed, it would mean that an initial step passed a
+    data-accessor (such as a side input / shuffle Iterable) down to the
+    next step in a stage.
+    """
+    pass
+
+  def __exit__(self, exc_type, exc_value, traceback):
+    """Exit the IO state."""
+    pass
+
+  def __enter__(self):
+    """Enter the IO state. This should track time spent blocked on IO."""
+    pass
+
+  def add_bytes_read(self, count):
+    pass
+
+
+class SideInputReadCounter(TransformIOCounter):
+  """Tracks time and bytes consumed while reading from side inputs.
+
+  This class is designed to track consumption of side inputs across fused steps.
+  We represent a side input as a declaring step, and an input index, which are
+  represented as an io_target.
+
+  The declaring step is the step that originally receives the side input for
+  consumption, and the input index in which the declaring step receives the side
+  input that we want to identify.
+
+  Note that the declaring step originally receives the side input, but it may
+  not be the only step that spends time reading from this side input.
+  """
+
+  def __init__(self, counter_factory, state_sampler, declaring_step,
+               input_index):
+    """Create a side input read counter.
+
+    Args:
+      counter_factory: A counters.CounterFactory to create byte counters.
+      state_sampler: A statesampler.StateSampler to transition into read states.
+      declaring_step: A string with the step name of the step that directly
+        receives the side input initially.
+      input_index: The index of the side input in the list of inputs of the
+        declaring step.
+
+    The side input is uniquely identified by (declaring_step, input_index);
+    where declaring_step is the step that receives the PCollectionView as a
+    side input, and input_index is the index of the PCView within the list of
+    inputs.
+    """
+    self._counter_factory = counter_factory
+    self._state_sampler = state_sampler
+    self.declaring_step = declaring_step
+    self.input_index = input_index
+
+    # Side inputs are set up within the start state of the first receiving
+    # state.
+    # We check the current state to create the internal counters.
+    self.check_step()
+
+  def check_step(self):
```

Comment:
<!--new_thread; commit:9c3a73fcac33b56a026e4ee575722c94c34cf64f; resolved:0-->
Do you think it could be clearer to rename this to something like `update_current_step()`?  (given how you expect this feature to progress)